Keep loading on the All Episodes tab even if items are filtered out

SebiderSushi opened the pull request:
Fixes #4414

SebiderSushi commented:
> How about adding a child function "hasMore()" and removing the state from the parent? This would avoid using null values (null values could lead to problems if someone does not read the method contract well enough).

Probably nicer than silently writing to the attribute.

SebiderSushi commented:
One remaining issue: I have a library of thousands of episodes. When i apply filter options that exclude most of the episodes is still have to scroll 150-episode pages at a time, regardless of how many episodes i actually get onto my screen due to the filter.

I think the best thing might be to redesign the filtering and apply filters directly at the database query. What do you think?

SebiderSushi commented:
okay i threw a bunch of extras in here, i guess the two latest commits are a little off topic for this specific pull request.

they each represent one dirty approach at fixing the UX when loading episodes from a library while filtering out most of it.
I'd welcome it, if the concept of filtering directly trough database queries (second commit) would be accepted and implemented cleanly in one way or the other.

ByteHamster commented:
Thanks for working on this. The bugfix looks good to me.

> okay i threw a bunch of extras in here, i guess the two latest commits are a little off topic for this specific pull request.

With the sql changes, we no longer need the original bugfix because the query already returns full pages. This makes me wonder if we even need java filtering code. If we rewrote the other filterable screens to use your new method, it would reduce the RAM footprint because we no longer need to create objects for items that are not displayed.

SebiderSushi commented:
> With the sql changes, we no longer need the original bugfix because the query already returns full pages. This makes me wonder if we even need java filtering code. If we rewrote the other filterable screens to use your new method, it would reduce the RAM footprint because we no longer need to create objects for items that are not displayed.

Hm, i guess it might be up for debate if querying a list from the database once and then applying several different filters might be more useful in terms of speed in some cases?

Otherwise i also think it could be a good idea to convert filtered lists to use filtered database queries. The intermediate third commit `Load new items on all episodes list until enough get filtered` will of couse be entirely obsolete in that scenario. I'd still leave the changes from the first two comments, tightly coupling the Database Query with the `data.size() < EPISODES_PER_PAGE` check to encapsulate it better. Alternatively, i'd concretely add a requirement to return a full sized page whenever possible to the documentation of `EpisodesListFragment.loadMoreData()`.

&nbsp;I am currently unsure where to do the conversion between the `FeedItemFilter` Object to the SQL Query, since both the filter object and the Database structure might change in the future while this conversion logic would have to stay in sync. Since the conversion havily depends on the database scheme & structure i veered towards the approach i implemented where it is tied to the `PodDBAdapter` class which already deals with `FeedItem` Objects and the likes anyway.

Going forward, i feel like encapsulating this more and decoupling it from consuming the internal String Array of `FeedItemFilter` would be better, e.g. by simply adding getters for the `FeedItemFiler` attributes. Also, i'd add a hint to the documentation of the `FeedItemFilter` mentioning that the DB Adapter should also be taken care of when changing the features of the Filter.  
  
  
  
If someone with database experience says that filtering via query doesn't have any performance gotchas or anything else and if the user experience will still be fine even while changing the filter options in rapid succession then i'd be happy to clean up my draft and search out some more places where lists could get filtered right from the database instead of in java.

ByteHamster commented:
> Hm, i guess it might be up for debate if querying a list from the database once and then applying several different filters might be more useful in terms of speed in some cases?

Currently, we load the full data again, even if we filter in Java. I think it would be more consistent (with fewer edge cases) to just continue loading the full data again. In that case, loading with SQL filters is more efficient.

> The intermediate third commit `Load new items on all episodes list until enough get filtered` will of couse be entirely obsolete in that scenario

I am not sure about that commit. The batched loading was added to prevent having to load many items at once (which takes time or fills the java heap), so the code increases loading time.

> Going forward, i feel like encapsulating this more and decoupling it from consuming the internal String Array of FeedItemFilter would be better, e.g. by simply adding getters for the FeedItemFiler attributes.

:+1: That sounds good. Because the filter attributes are never changed, I think it would also be possible to just make them `public final` instead of adding getters.

> If someone with database experience says that filtering via query doesn't have any performance gotchas or anything else and if the user experience will still be fine even while changing the filter options in rapid succession then i'd be happy to clean up my draft and search out some more places where lists could get filtered right from the database instead of in java.

I think it should be good. If we did notice performance problems later, we can always just change the query to optimize things (eg `JOIN` instead of subqueries - but I think it should be OK for now).

SebiderSushi commented:
Okay i think only the `AllEpisodes` List and `FeedItemList` are actually filtered and i didn't quite want to touch the `FeedItemListFragment` because it seems very convoluted there. Also it seems to me that `DBReader.loadAdditionalFeedItemListData()` could be replaced with some `JOIN`s when moving to SQL filtering.

I can conform to the Checkstyle rules if desired but i really think that the cascade of if statements profits in readability if each of them fits onto one line.

ByteHamster commented:
> Also it seems to me that DBReader.loadAdditionalFeedItemListData() could be replaced with some JOINs when moving to SQL filtering.

When loading the `Feed` objects with joins, each feed item has its own feed instance. That consumes a lot of RAM unnecessarily. Currently, all feed items of the same feed share their feed object.

Queue and favorites can probably be done more efficiently with left joins than what is currently done.

ByteHamster commented on core/src/main/java/de/danoeh/antennapod/core/feed/FeedItemFilter.java:
```diff
@@ -78,6 +80,45 @@ public FeedItemFilter(String[] properties) {
+        String tableFavorites = PodDBAdapter.TABLE_NAME_FAVORITES;
+
+        List<String> statements = new ArrayList<>();
+        if (showPlayed)        statements.add(keyRead + " = 1 ");
```
> I can conform to the Checkstyle rules if desired but i really think that the cascade of if statements profits in readability if each of them fits onto one line.

Readability was probably the cause of the [goto fail bug](https://en.wikipedia.org/wiki/Unreachable_code#goto_fail_bug), too. I would prefer to always use braces, even if it makes the code a bit longer.

ByteHamster commented:
Hi @SebiderSushi are you still interested in this PR? It would make a great addition to AntennaPod because it makes filtering and loading items a lot more efficient (and fixes #4414 at the same time).

ByteHamster commented:
Thanks for working on this and sorry for the long delay before finally merging.

SebiderSushi commented:
Sorry for neglecting that patch so badly :(
Thank you for prepping it up and tying up the loose ends!
